spa/plugins/bluez5/a2dp-source.c

```diff
@@ -0,0 +1,1064 @@
+
+	this = SPA_CONTAINER_OF(node, struct impl, node);
+
+	this->callbacks = callbacks;
+	this->callbacks_data = data;
+
+	return 0;
+}
+
+static int impl_node_add_port(struct spa_node *node, enum spa_direction direction, uint32_t port_id,
+		const struct spa_dict *props)
+{
+	return -ENOTSUP;
+}
+
+static int impl_node_remove_port(struct spa_node *node, enum spa_direction direction, uint32_t port_id)
+{
+	return -ENOTSUP;
+}
+
+static int
+impl_node_port_enum_params(struct spa_node *node, int seq,
+			enum spa_direction direction, uint32_t port_id,
+			uint32_t id, uint32_t start, uint32_t num,
+			const struct spa_pod *filter)
+{
+
+	struct impl *this;
+	struct spa_pod *param;
+	struct spa_pod_builder b = { 0 };
+	uint8_t buffer[1024];
+	struct spa_result_node_params result;
+	uint32_t count = 0;
+
+	spa_return_val_if_fail(node != NULL, -EINVAL);
+	spa_return_val_if_fail(num != 0, -EINVAL);
+
+	this = SPA_CONTAINER_OF(node, struct impl, node);
+
+	result.id = id;
+	result.next = start;
+      next:
+        result.index = result.next++;
+
+	spa_pod_builder_init(&b, buffer, sizeof(buffer));
+
+	switch (id) {
+	case SPA_PARAM_EnumFormat:
+		if (result.index > 0)
+			return 0;
+
+		switch (this->transport->codec) {
+		case A2DP_CODEC_SBC:
+		{
+			a2dp_sbc_t *config = this->transport->configuration;
+			struct spa_audio_info_raw info = { 0, };
+
+			info.format = SPA_AUDIO_FORMAT_S16;
+			if ((info.rate = a2dp_sbc_get_frequency(config)) < 0)
+				return -EIO;
+			if ((info.channels = a2dp_sbc_get_channels(config)) < 0)
+				return -EIO;
+
+			switch (info.channels) {
+			case 1:
+				info.position[0] = SPA_AUDIO_CHANNEL_MONO;
+				break;
+			case 2:
+				info.position[0] = SPA_AUDIO_CHANNEL_FL;
+				info.position[1] = SPA_AUDIO_CHANNEL_FR;
+				break;
+			default:
+				return -EIO;
+			}
+
+			param = spa_format_audio_raw_build(&b, id, &info);
+			break;
+		}
+		case A2DP_CODEC_MPEG24:
+		{
+			/* not implemented yet */
+			spa_log_error(this->log, "a2dp mpeg24 codec not implemented yet");
+			return -EIO;
+		}
+		default:
+			return -EIO;
+		}
+		break;
+
+	case SPA_PARAM_Format:
+		if (!this->have_format)
+			return -EIO;
+		if (result.index > 0)
+			return 0;
+
+		param = spa_format_audio_raw_build(&b, id, &this->current_format.info.raw);
+		break;
+
+	case SPA_PARAM_Buffers:
+		if (!this->have_format)
+			return -EIO;
+		if (result.index > 0)
+			return 0;
+
+		param = spa_pod_builder_add_object(&b,
+			SPA_TYPE_OBJECT_ParamBuffers, id,
+			/* 8 buffers are enough to asure we always have one available buffer when decoding */
+			SPA_PARAM_BUFFERS_buffers, SPA_POD_CHOICE_RANGE_Int(8, 8, MAX_BUFFERS),
+			SPA_PARAM_BUFFERS_blocks,  SPA_POD_Int(1),
+			/* looks like uncompressed SBC data is 5 times bigger than compressed SBC data, so make
+			 * sure the buffer has enought space for it by making it 6 times bigger */
+			SPA_PARAM_BUFFERS_size,    SPA_POD_Int(this->props.min_latency * this->frame_size * 6),
```

Comment:
This sounds a bit wrong. this->frame_size refers to the raw audio frame size, so in theory, multiplying this with the latency **should** give out the size of the raw audio buffer.

I guess the real problem is that the remote end (the phone) does not know what is our configured min_latency, so it picks its own latency, which in your case is probably min_latency * 5 = 128 * 5 = 640 samples. The proper solution would be to negotiate that (i have no idea how). Another solution would be to use the max_latency instead of the min_latency, which is 1024 samples. Theoretically, even if we negotiate, we should never use a latency bigger than max_latency, so max_latency * frame_size should be the maximum space you will ever need.

Reply:
Yes, the phone does not know the configured latency and it picks up the minimum latency. I have changed it to use the max latency until we find a better solution